### DISCUSSION TITLE
ImGui in two windows?

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1317] [General]
I'm using ImGuizmo, an ImGui based library, for manipulating mesh transformation matrices. My application is using a tab interface where there is a vsg::window in each tab.

However, when trying to create a second tab, my program crashed with the error:

```
.../vsgimgui-src/src/imgui/backends/imgui_impl_vulkan.cpp:1101: bool ImGui_ImplVulkan_Init(ImGui_ImplVulkan_InitInfo*): Assertion `io.BackendRendererUserData == nullptr && "Already initialized a renderer backend!"' failed.
```
This error was received when doing:
```
auto window = vsg::Window::create()
auto myCommands = vsg::ImGui::Commands::create(...);
auto renderImGui = vsgImGui::RenderImGui::create(window, myCommands);
```

Looking at the source code, indeed confirms that there is a call to `ImGui_ImplVulkan_Init()` in this call chain, which prevents calling `vsgImGui::RenderImGui::create();` for more than one window. 

Is there any workaround for this, or this a limitation in ImGui?

**Comment by robertosfield** [comment]
I don' know if ImGui is capable of working within multiple windows.  I haven't used ImGui much myself, my experience is has been  refining community contributions to vsgImGui rather than as a ImGui user.  I've leaned enough to get things working for the single window usage case.

Have a look at the ImGui project decs to see if there is a multi-window example.  If there is perhaps vsgImGui can be refactored to handle the multi-window usage case.

**Reply by dov** [reply]
I found it! You need to save the Imgui context for each window, and set it before doing any imgui ops for that window. You can already do it through Imgui calls, so there is no need for additional support in vsgImGui.

```c++
// for each window
_imGuiContext = ImGui::CreateContext();
:
// before doing any img ui calls
ImGui::SetCurrentContext(_imGuiContext);
```
